pkg/solana/txm/txm.go

```diff
@@ -377,139 +399,206 @@ func (txm *Txm) confirm() {
+	}
+}
+
+// handleProcessedSignatureStatus handles the case where a transaction signature is in the "processed" state on-chain.
+// It updates the transaction state in the local memory and checks if the confirmation timeout has been exceeded.
+// If the timeout is exceeded, it marks the transaction as errored.
+func (txm *Txm) handleProcessedSignatureStatus(sig solanaGo.Signature) {
+	// update transaction state in local memory
+	id, err := txm.txs.OnProcessed(sig)
+	if err != nil && !errors.Is(err, ErrAlreadyInExpectedState) {
+		txm.lggr.Errorw("failed to mark transaction as processed", "signature", sig, "error", err)
+	} else if err == nil {
+		txm.lggr.Debugw("marking transaction as processed", "id", id, "signature", sig)
+	}
+	// check confirm timeout exceeded if TxConfirmTimeout set
+	if txm.cfg.TxConfirmTimeout() != 0*time.Second && txm.txs.Expired(sig, txm.cfg.TxConfirmTimeout()) {
+		id, err := txm.txs.OnError(sig, txm.cfg.TxRetentionTimeout(), Errored, TxFailDrop)
+		if err != nil {
+			txm.lggr.Infow("failed to mark transaction as errored", "id", id, "signature", sig, "timeoutSeconds", txm.cfg.TxConfirmTimeout(), "error", err)
+		} else {
+			txm.lggr.Debugw("tx failed to move beyond 'processed' within confirm timeout", "id", id, "signature", sig, "timeoutSeconds", txm.cfg.TxConfirmTimeout())
+		}
+	}
+}
+
+// handleConfirmedSignatureStatus handles the case where a transaction signature is in the "confirmed" state on-chain.
+// It updates the transaction state in the local memory.
+func (txm *Txm) handleConfirmedSignatureStatus(sig solanaGo.Signature) {
+	id, err := txm.txs.OnConfirmed(sig)
+	if err != nil && !errors.Is(err, ErrAlreadyInExpectedState) {
+		txm.lggr.Errorw("failed to mark transaction as confirmed", "id", id, "signature", sig, "error", err)
+	} else if err == nil {
+		txm.lggr.Debugw("marking transaction as confirmed", "id", id, "signature", sig)
+	}
+}
+
+// handleFinalizedSignatureStatus handles the case where a transaction signature is in the "finalized" state on-chain.
+// It updates the transaction state in the local memory.
+func (txm *Txm) handleFinalizedSignatureStatus(sig solanaGo.Signature) {
+	id, err := txm.txs.OnFinalized(sig, txm.cfg.TxRetentionTimeout())
+	if err != nil {
+		txm.lggr.Errorw("failed to mark transaction as finalized", "id", id, "signature", sig, "error", err)
+	} else {
+		txm.lggr.Debugw("marking transaction as finalized", "id", id, "signature", sig)
+	}
+}
+
+// rebroadcastExpiredTxs attempts to rebroadcast all transactions that are in broadcasted state and have expired.
+// An expired tx is one where it's blockhash lastValidBlockHeight (last valid block number) is smaller than the current block height (block number).
+// If any error occurs during rebroadcast attempt, they are discarded, and the function continues with the next transaction.
+func (txm *Txm) rebroadcastExpiredTxs(ctx context.Context, client client.ReaderWriter) {
```

Comment:
this feels like there's a potential race condition here

* tx is broadcast to the network and is included just before blockhash expiration
* confirmer does not see it yet because it waits for `confirmed` commitment level
* TXM marks a base transaction as expired because the block threshold has passed 
* deletes the saved tx id from the in-memory storage, and rebroadcasts a tx with a new blockhash
* this could go on forever?

Reply:
Wouldn't this be handled gracefully by the existing code? 

The `processConfirmations` method is called before `rebroadcastExpiredTxs` updating each signature status. Here we would detect if a transaction [transitioned from Broadcasted to Processed](https://github.com/smartcontractkit/chainlink-solana/blob/b4b4fd54f8999f83f66f194bb48305499c511a8f/pkg/solana/txm/txm.go#L467-L470).

Then inside `rebroadcastExpiredTxs` we call `ListAllExpiredBroadcastedTxs`, [where we only consider the ones which have `Broadcasted` state](https://github.com/smartcontractkit/chainlink-solana/blob/b4b4fd54f8999f83f66f194bb48305499c511a8f/pkg/solana/txm/pendingtx.go#L230C1-L241C1)

Reply:
I think we would avoid that scenario since we only call `rebroadcastExpiredTxs` after processing the latest confirmation status changes. Since this method only processes `Broadcasted` transaction, the confirmation logic would handle any last minute state changes before we assess transaction for expiry. So in your scenario, we would notice that the transaction advanced to `Processed`, mark it accordingly in our in-memory storage, then move on to the expiry check. The expiry check would notice there aren't any transactions in `Broadcasted` state that are expired and just move on.

Reply:
We would advance the state from non-existent (internally what we consider `Broadcasted`) to `Confirmed` and also avoid rebroadcasting with a new hash in that case. Our state transition methods don't require us to move one step up at a time in case of slow polling or other reasons. We can even jump straight to `Finalized`. We only delete all previous attempts/sigs if we can confirm that the tx has a non-existent state and its blockhash has expired. I was thinking it was better to clean up those old attempts/sigs than keep them around and waste cycles on status checks on them since they're expired anyways.

Reply:
ok - the chances of this race happening might be low

is there logging around a transaction signature that doesn't match any tx id? something that fires if an assumption is broken?

Reply:
If one of the state transition methods gets a sig that doesn't match to an ID in the internal map, it should return a `ErrSigDoesNotExist` error. Depending where the method was called, we have some logging to surface an assumption violation. But not in all cases like the confirmation expiration where we can maybe assume a different process cleaned up the signatures.